compiler/generator_rust.c

```diff
@@ -0,0 +1,1347 @@
+
+    write_comment(g, p);
+    g->S[0] = p->mode == m_forward ? "" : "_b";
+    g->V[0] = p->name;
+    write_failure_if(g, "!env.eq_s~S0(&~V0)", p);
+}
+
+static void generate_literalstring(struct generator * g, struct node * p) {
+
+    symbol * b = p->literalstring;
+    write_comment(g, p);
+    g->S[0] = p->mode == m_forward ? "" : "_b";
+    g->L[0] = b;
+    write_failure_if(g, "!env.eq_s~S0(&~L0)", p);
+}
+
+static void generate_setup_context(struct generator * g) {
+  struct name * q;
+  w(g, "~Mlet mut context = &mut Context {~+~N");
+  for (q = g->analyser->names; q; q = q->next) {
+    g->V[0] = q;
+    switch (q->type) {
+    case t_string:
+      w(g, "~M~W0: String::new(),~N");
+      break;
+    case t_integer:
+      w(g, "~M~W0: 0,~N");
+      break;
+    case t_boolean:
+      w(g, "~M~W0: false,~N");
+      break;
+    }
+  }
+  w(g, "~-~M};~N");
+}
+
+
+static void generate_define(struct generator * g, struct node * p) {
+
+    struct name * q = p->name;
+
+    struct str * saved_output = g->outbuf;
+
+    g->V[0] = q;
+
+    if (q->type == t_routine) {
+        w(g, "~N~Mfn ~W0(env: &mut SnowballEnv, context: &mut Context) -> bool {~+~N");
+    } else {
+        w(g, "~N~Mpub fn ~W0(env: &mut SnowballEnv) -> bool {~+~N");
+        generate_setup_context(g);
+    }
+    // w(g, "~Mprintln!(\"~W0: \\t\\t\\t{:?}\", env);~N");
+
+    g->outbuf = str_new();
+
+    g->next_label = 0;
+    g->var_number = 0;
+
+    str_clear(g->failure_str);
+    g->failure_label = x_return;
+    g->unreachable = false;
+    generate(g, p->left);
+    if (!g->unreachable) w(g, "~Mreturn true;~N");
+    w(g, "~-~M}~N");
+
+    str_append(saved_output, g->outbuf);
+    str_delete(g->outbuf);
+    g->outbuf = saved_output;
+}
+
+static void generate_substring(struct generator * g, struct node * p) {
+
+    struct among * x = p->among;
+
+    write_comment(g, p);
+
+    g->S[0] = p->mode == m_forward ? "" : "_b";
+    g->I[0] = x->number;
+
+    if (x->command_count == 0 && x->starter == 0) {
+        write_failure_if(g, "env.find_among~S0(~A_~I0, context) == 0", p);
+    } else {
+        writef(g, "~Mlet among_var = env.find_among~S0(~A_~I0, context);~N", p);
+        write_failure_if(g, "among_var == 0", p);
+    }
+}
+
+static void generate_among(struct generator * g, struct node * p) {
+
+    struct among * x = p->among;
+    int case_number = 1;
+
+    if (x->substring == 0) generate_substring(g, p);
+    if (x->command_count == 0 && x->starter == 0) return;
+
+    if (x->starter != 0) generate(g, x->starter);
+
+    p = p->left;
+    if (p != 0 && p->type != c_literalstring) p = p->right;
+    w(g, "~Mif among_var == 0 {~N~+");
+    write_failure(g);
+    g->unreachable = false;
+    w(g, "~-~M} ");
+
+    while (p != 0) {
+        if (p->type == c_bra && p->left != 0) {
+            g->I[0] = case_number++;
+            w(g, "else if among_var == ~I0 {~N~+");
+            generate(g, p);
+            w(g, "~-~M} ");
```

Comment:
This leaves a line-final space after the last among clause. Instead of adding spaces after the braces here and on line 1074, add a space before the `else` on line 1079.